internal/cloudsql/metadataexchange.go

```diff
@@ -0,0 +1,206 @@
+	return c.res
+}
+
+// Read implements io.Reader interface, delegating to the internal buffered reader.
+func (c *MDXConn) Read(b []byte) (int, error) {
+	if !c.firstRead {
+		res, mdxErr := c.readMDX()
+		if mdxErr != nil {
+			return 0, mdxErr
+		}
+		c.res = res
+		c.firstRead = true
+	}
+	return c.r.Read(b)
+}
+
+// Read implements io.Reader interface, delegating to the internal buffered reader.
+func (c *MDXConn) Write(b []byte) (int, error) {
+	if !c.firstWrite {
+		if c.req != nil {
+			mdxErr := c.writeMDX(c.req)
+			if mdxErr != nil {
+				return 0, mdxErr
+			}
+		}
+		c.firstWrite = true
+	}
+	return c.w.Write(b)
+}
+
+// readMDX reads the MDX request message from the socket.
+func (c *MDXConn) readMDX() (*mdx.MetadataExchangeResponse, error) {
+	// ctx only used for debug logging
+	ctx := context.Background()
+	c.logger.Debugf(ctx, "[%v] Reading MDX response...", c.cn)
+
+	// Progressively peek from 1 to 8 bytes from the socket. If at any point it does not match, there's an i/o
+	// error, or if the read times out, then return.
+	for l := 1; l <= mdxSignatureLength; l++ {
+		peekBuf, err := c.r.Peek(l)
+		if errors.Is(err, os.ErrDeadlineExceeded) {
+			c.logger.Debugf(ctx, "MDX Read Timeout Error: error peeking first %d bytes for the MDX request: %v", l, err)
+			// Ignore IO timeout error. Assume that if no MDX was sent within 250ms following the
+			// CLIENT_PROTOCOL_TLS handshake, then there will be no MDX request, and the database client is waiting for
+			// the server to respond.
+			return nil, errtype.NewDialError(fmt.Sprintf("MDX Read Timeout Error: error peeking first %d bytes for the MDX request: %v", l, err), c.cn, err)
+		} else if err != nil {
+			c.logger.Debugf(ctx, "MDX Read Error: error peeking first %d bytes for the MDX request: %v", l, err)
+			return nil, errtype.NewDialError(fmt.Sprintf("MDX Read Error: error peeking first %d bytes for the MDX request", l), c.cn, err)
+		}
+
+		if !bytes.Equal(peekBuf, signature[:len(peekBuf)]) {
+			c.logger.Debugf(ctx, "MDX Read: No MDX response, first %d bytes do not match MDX signature. %v", l)
+			return nil, nil
+		}
+	}
+
+	// Received a MDX signature, will read the response.
+	// Read the 12 byte mdx header from the stream
+	mdxHeaderBytes := make([]byte, mdxHeaderLength)
+	_, err := io.ReadFull(c.r, mdxHeaderBytes)
+
+	if errors.Is(err, os.ErrDeadlineExceeded) {
+		return nil, errtype.NewDialError("MDX Read Timeout: read response header", c.cn, err)
+	} else if err != nil {
+		return nil, errtype.NewDialError("MDX Read Error: read response header", c.cn, err)
+	}
+
+	// Get the MDX message length from the MDX header
+	msgLen := binary.BigEndian.Uint32(mdxHeaderBytes[mdxSignatureLength:mdxHeaderLength])
+	if msgLen > maxMessageSize {
+		return nil, errtype.NewDialError(fmt.Sprintf("MDX Read Format: Response too long, %d bytes", msgLen), c.cn, err)
+	}
+
+	// Read the MDX message
+	mdxBytes := make([]byte, msgLen)
```

Comment:
To avoid the performance tax of heap allocation, we use a sync.Pool in the AlloyDB Go Connector.

See https://github.com/GoogleCloudPlatform/alloydb-go-connector/blob/227c2d92167aa9c62ed904588ce202f71f37fc7c/dialer.go#L501-L502 and https://github.com/GoogleCloudPlatform/alloydb-go-connector/blob/227c2d92167aa9c62ed904588ce202f71f37fc7c/dialer.go#L554-L577